### PR TITLE
Remove casts to imported anonymous types and add a validation to disallow the same

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title, as a temporary fix for https://github.com/ballerina-platform/ballerina-lang/issues/28262.

Not sure if we can write a test for either change.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -4435,8 +4438,11 @@ private BLangBlockStmt desugarForeachToWhile(BLangForeach foreach, BLangSimpleVa
         // However, we are within the while loop. hence the $result$ can never be nil nor error.
         // Therefore cast $result$ to non-nilable type. i.e `int item = <>$result$.value;`
         BLangFieldBasedAccess valueAccessExpr = getValueAccessExpression(foreach.pos, foreach.varType, resultSymbol);
-        valueAccessExpr.expr = addConversionExprIfRequired(valueAccessExpr.expr,
-                types.getSafeType(valueAccessExpr.expr.type, true, false));
+
+        BLangExpression expr = valueAccessExpr.expr;
+        // Since `$result$` is always a record (with a single `value` field), add a cast to `map<any|error>`
+        // to avoid casting to an anonymous type - https://github.com/ballerina-platform/ballerina-lang/issues/28262.
+        valueAccessExpr.expr = addConversionExprIfRequired(expr, symTable.mapAllType);
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:4445]: Out of curiosity, what happens if we don't have this conversion here? Issues in jvm codegen or does it get re-added somehow via desugaring the desugared constructs?

**Review by rdhananjaya — APPROVED** [review]

**Review by KRVPerera — APPROVED** [review]
Better if we can add a positive test case which was failing previously with anonemous type cast.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 1 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:4445]: So the type of `expr` (`$result$`) here is the union of a record and nil - `record {|T value;|}?`. Although semantically it is invalid to do `$result$.value` in such a scenario, desugaring proceeds as a field access on a record union due to the check at https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java#L5171. And then at org.wso2.ballerinalang.compiler.bir.codegen.JvmInstructionGen#generateMapLoadIns we don't seem to be adding a cast so probably works as is there?

**Comment by MaryamZi** [comment]
> Better if we can add a positive test case which was failing previously with anonemous type cast.

Any suggestions on how we can add such a test? 

There are several conditions that needs to be met. For example if we consider using `foreach` with an array
1. need to have two compatible lang.array versions where both have different names for the anonymous types
2. need to build the JAR using the pack with one version and then run it using a pack with the other

**Comment by MaryamZi** [comment]
> > Better if we can add a positive test case which was failing previously with anonemous type cast.
> 
> Any suggestions on how we can add such a test?
> 
> There are several conditions that needs to be met. For example if we consider using `foreach` with an array
> 
> 1. need to have two compatible lang.array versions where both have different names for the anonymous types
> 2. need to build the JAR using the pack with one version and then run it using a pack with the other

Will go ahead and merge this for now, will add a test in a separate PR if we can figure out a way to test this.

**Pull request merged**